katsdpingest/ingest_session.py

```diff
@@ -937,50 +979,63 @@ def _init_ig_sd(self) -> None:
-        self.rx_spead_ibv = args.cbf_ibv
-        self.rx_spead_max_streams = args.input_streams
-        self.rx_spead_max_packet_size = args.input_max_packet_size
-        self.rx_spead_buffer_size = args.input_buffer
-        self.sd_spead_rate = args.sd_spead_rate / args.clock_ratio if args.clock_ratio else 0.0
+        self.rx_spead_ibv: bool = args.cbf_ibv
+        self.rx_spead_max_streams: int = args.input_streams
+        self.rx_spead_max_packet_size: int = args.input_max_packet_size
+        self.rx_spead_buffer_size: int = args.input_buffer
+        self.sd_spead_rate: float = (
+            args.sd_spead_rate / args.clock_ratio if args.clock_ratio else 0.0
+        )
         self.sd_spead_ifaddr = katsdpservices.get_interface_address(args.sdisp_interface)
         self.channel_ranges = channel_ranges
-        self.telstate = telstate
-        self.telstate_cbf = utils.cbf_telstate_view(telstate, args.cbf_name)
-        self.telstate_sdisp = telstate.view('sdp', exclusive=True).view(args.l0_spectral_name)
-        self.cbf_attr = cbf_attr
-        self.src_stream = args.cbf_name
-        self.use_data_suspect = args.use_data_suspect
-
-        self._init_baselines(args.antenna_mask)
+        self.telstate = telstate_cbf.root()
+        self.telstate_cbf = telstate_cbf
+        self.telstate_sdisp = self.telstate.view('sdp', exclusive=True).view(args.l0_spectral_name)
+        self.cbf_attr = system_attrs.cbf_attr
+        self.src_stream: str = args.cbf_name
+        self.use_data_suspect: bool = args.use_data_suspect
+        self._telstate_updates: List[Tuple[str, str, Any]] = []   # Values to populate in telstate
+
+        self._init_baselines(system_attrs)
         self._init_time_averaging(args.output_int_time, args.sd_int_time)
         self._init_sensors(my_sensors)
-        self._init_tx(args)  # Note: must be run after _init_time_averaging, before _init_proc
+        self._init_tx(args)  # NB: must be run after _init_time_averaging, before _init_proc
         self._init_proc(context, args.excise, 'cont' in self.tx)
         self._init_resources()
 
         # Instantiation of input streams is delayed until the asynchronous task
         # is running, to avoid receiving data we're not yet ready for.
-        self.rx = None       # type: Optional[receiver.Receiver]
+        self.rx: Optional[receiver.Receiver] = None
         self._init_ig_sd()
 
         # Record information about the processing in telstate
         if args.name is not None:
             descriptions = _fix_descriptions(list(self.proc.descriptions()))
-            process_view = self.telstate.view(args.name.replace('.', '_'))
-            utils.set_telstate_entry(process_view, 'process_log', descriptions)
+            process_prefix = args.name.replace('.', '_')
+            self._telstate_updates.append((process_prefix, 'process_log', descriptions))
+
+    async def populate_telstate(self) -> None:
```

Comment:
I had to pull a similar stunt in `CaptureSession`, where I collect obs params and script logs before `capture_init` and then send them to telstate afterwards. I toyed with the idea of a `CachingTelescopeState` to serve both our needs, but probably too complicated (for one, the part of getting the real telstate is trickier).